README.md
# NewsArticleGen

## Overview
NewsArticleGen is an AI-powered tool designed to scan web news and generate preliminary article ideas for professionals. The system analyzes current news, creates topic suggestions across various categories, and matches these ideas with suitable professional profiles, streamlining the content creation process.

## Features
- **News Analysis**: Scans and processes news articles from various sources
- **Idea Generation**: Creates article ideas across multiple categories including International, National, Economy, Science & Technology, Health, Environment, Crime & Justice, Education, and Social & Humanitarian topics
- **Professional Matching**: Uses embeddings to find the most suitable professionals for each topic
- **Article Generation**: Creates preliminary article drafts based on selected ideas and professional profiles
- **Multilingual Support**: Processes and generates content in Spanish (with potential for other languages)

## Project Structure
- `create_embeddings.py`: Generates embeddings for professional profiles
- `find_professional.py`: Matches topics with appropriate professionals
- `generate_ideas.py`: Creates article ideas from news content
- `generate_article.py`: Produces preliminary article drafts
- `python_dump.py`: Contains utility functions used throughout the project
- `myutils/embeddings_utils.py`: Provides embedding-related utility functions

## Installation

### Prerequisites
- Python 3.8+
- OpenAI API key

### Setup
1. Clone the repository
```bash
git clone https://github.com/yourusername/newsarticlegen.git
cd newsarticlegen
```

2. Install dependencies
```bash
pip install -r requirements.txt
```

3. Set up your OpenAI API key as an environment variable
```bash
export OPENAI_API_KEY='your-api-key'
```

4. Create the necessary directories
```bash
mkdir -p resources/raw_profiles resources/embedded_profiles
```

## Usage

### 1. Prepare professional profiles
Add professional profile text files to the `resources/raw_profiles` directory.

### 2. Generate embeddings for profiles
```bash
python create_embeddings.py
```

### 3. Generate article ideas from news
- Add news content to `resources/News.txt`
- Run the idea generator:
```bash
python generate_ideas.py
```

### 4. Find professionals for specific topics
- Add your topic to `resources/User_prompt_find_professional.txt`
- Run:
```bash
python find_professional.py
```

### 5. Generate articles
- Set up your article request in `resources/User_prompt_generate_article.txt`
- Run:
```bash
python generate_article.py
```

## Customization
- Modify the prompts in the Python files to adjust the style and focus of generated content
- Adjust the number of ideas per category in `generate_ideas.py`
- Change the matching criteria in `find_professional.py`

## Contributing
Contributions are welcome! Please feel free to submit a Pull Request.

## License
[Specify your license here]

## Acknowledgments
- Built using OpenAI's API
- Uses embedding techniques for semantic matching


resources/News.txt
Clarin.

"Como una final": los hinchas argentinos no se achicaron y se hicieron sentir en el clásico contra Brasil

Fueron entre 2.500 y 3.000 que pudieron comprar su ticket con documento argentino para presenciar el duelo por las Eliminatorias.
En las playas, en las calles y en las tribunas se vivió el clásico, pero sin ningún tipo de problema.
Es la sexta fecha de las Eliminatorias Sudamericanas rumbo al Mundial 2026, pero si se necesitaría un ejemplo para algún out of context fútbol de lo que significa un clásico entre Brasil y Argentina habría que mostrarle las horas previas, el partido y el post de este martes en Río de Janeiro.
Lejos de la violencia que rodeó la final de la Copa Libertadores entre Boca y Fluminense hace tres semanas, el duelo entre las selecciones más importantes de Sudamérica se vivió intensamente en las playas, en las calles y en las tribunas del mítico estadio, con sus 70.000 entradas agotadas desde hace un par de días atrás. De esos, entre 2.500 y 3.000 pudieron comprar su ticket como hinchas argentinos, pero a ellos se le sumaron los que viven en Brasil y accedieron a su lugar en el Maracaná.
Fue Copacabana, en el posto 5, el centro neurálgico para los hinchas albicelestes. Recorriendo historias se encontraba de todo. Estaban los más organizados bajo el ala de la 'Banda Argentina' (@bandargentina en Instagram), unas 40 o 50 personas que lo previeron desde hace varios meses después de crear un grupo de WhatsApp en el Mundial, del que hoy ya forman parte más de 300 fanáticos de la Scaloneta. Ellos se plantaron en el centro de la tribuna Sur y se hicieron sentir, pese a estar en desventaja, además de preparar cotillón para la salida de los equipos.


Infobae

“The Marvels” no despega en taquilla y es el peor estreno de Marvel este 2023

El trío de heroínas protagonizado por Brie Larson, Iman Vellani y Teyonah Parris no recaudó las cifras esperadas por Disney
The Marvels, la más reciente entrega del Universo Cinematográfico de Marvel, ha enfrentado un desafiante fin de semana de apertura, según datos de Box Office Mojo. La película, que entrelaza los poderes y las historias de Carol Danvers (Brie Larson), Kamala Khan (Iman Vellani) y Monica Rambeau (Teyonah Parris), fue distribuida por Walt Disney Studios Motion Pictures y se estrenó el pasado 9 de noviembre en Latinoamérica y el 10 de noviembre en los Estados Unidos. Bajo el género de acción, aventura y fantasía, y con una clasificación PG-13, la película recaudó 47 millones de dólares en territorio estadounidense y 41.5 millones a nivel internacional, sumando un total de 88.5 millones a nivel mundial.



La nacion 

En la Argentina no existen los medios públicos

Las organizaciones de comunicación estatal siempre fueron usadas con intereses facciosos; el dilema que plantean los cambios tecnológicos y culturales

La expresión medio público refiere a una organización de comunicación que responde de manera directa a los intereses de los ciudadanos en su conjunto. El ejemplo paradigmático es la BBC de Londres, que cumplió 100 años en 2022, fue creada por un decreto de la Corona, y en la que el gobierno británico tiene prohibido hacer propaganda.

En Argentina nunca hubo medios públicos sino organizaciones de comunicación propagandísticas al servicio del poder de turno tanto en gobiernos democráticos como dictatoriales. Tampoco hubo intención real de tener medios verdaderamente públicos y la sociedad no parece haberlos reclamado. A juzgar por los niveles de audiencia, los ciudadanos argentinos parecen haberse distanciado hace tiempo de estas emisoras.
Durante los cuatro gobiernos kirchneristas, sobre todo luego de la sanción de la ley de medios en 2009, los “medios públicos” adquirieron características facciosas y se volvieron una herramienta de persecución ideológica. En la cosmovisión populista, público se equiparó a pueblo, y este último a los votantes del peronismo en sus diferentes marcas: Frente para la Victoria, Frente de Todos o Unión por la Patria.

Durante la campaña, el entonces candidato de La Libertad Avanza, Javier Milei, prometió privatizar Radio Nacional, Canal 7 y la agencia Télam. En sus primeras declaraciones como presidente electo, Milei confirmó que buscará vender esas empresas, que en términos mediáticos casi no tienen valor, pero cuyos inmuebles y propiedades podrían despertar interés de posibles inversores.

Ante esta confirmación, directivos políticos -ente ellos las presidentas de RTA, Rosario Lufrano, y la de Télam, Bernarda Llorente, y sindicalistas -entre ellos Jorge Batata Rodríguez, del Sipreba- enumeraron supuestas razones por las que el primer mandatario electo debería revisar su decisión. Sin embargo, esos argumentos son fáciles de rebatir por falaces, extemporáneos y directamente mentirosos.
Un comunicado formal de Lufrano, Llorente y otros funcionarios políticos designados por el gobierno saliente, sintetiza esas ideas en 10 puntos que podrían agruparse a su vez en cuatro ideas:

Carácter no comercial de los medios estatales y su producto, la información y los contenidos formativos.
Pluralismo, diversidad e inclusión de agendas alternativas.
Alcance a todo el territorio, “soberanía informativa y cultural” y la difusión de la Argentina al mundo.
Transparencia y rendición de cuentas a la ciudadanía y órganos de control.
El supuesto carácter no comercial se contradice con la recurrente -y en general fallida- apuesta de la emisora a competir con el sector privado, en especial en la televisación de grandes eventos. Competencia que termina siendo desleal porque el Estado parece no asumir riesgos y porque dicta normas y leyes que sólo aplica a privados y no a sí mismo. También es contradictorio con los conflictos de interés que cruzan la producción de contenidos, frecuentemente tercerizados en productoras privadas de empresarios políticamente afines al gobierno. Este argumento es falaz.


generate_ideas.py
import json

from openai import OpenAI
import os
import numpy as np
import pandas as pd
from ast import literal_eval
from myutils.embeddings_utils import get_embedding, cosine_similarity

# start timer
import time

from python_dump import read_embedding, parse_json, search_reviews, read_file, load_embeddings

start = time.time()

key = os.environ.get('OPENAI_API_KEY')

client = OpenAI(api_key=key)

prompt = """
             Agarrá las notiticias que te van a venir a continuación y generá a partir de las siguientes categorías ideas para columnas de opinión.
             Categorías:
                Internacional
                Nacional
                Economía
                Ciencia y Tecnología
                Salud
                Medio Ambiente
                Crimen y Justicia
                Educación
                Social y Humanitario
             Generá 5 ideas por categoría.
             Las ideas no tienen que estar directamente relacionadas con la noticia, pueden ser ideas que se te ocurran a partir de la noticia o un analisis de una nueva perpectiva.
             Las ideas deben incluir una lista de temas a los que estén relacionadas. 
             Las ideas no deben incluir temas politicos a menos que sea una noticia politica.
             Valorá más ideas que no incluyan temas politicos.
             2 de las 5 ideas deben ser un poco controversiales.
             Preferí ideas que puedan ser desarrolladas en una columna de opinión en un aspecto más profesional.
             Las ideas tienen que estar en español.
             Escribí las ideas en formato de json, por ejemplo:
             [
                {
                    "categoria": "Internacional",
                    "ideas": [
                        {
                            "idea": "idea1",
                            "temas": [tema1, tema2, tema3],
                        }
                        {
                            "idea": "idea2",
                            "temas": [tema1, tema2, tema3],
                        }
                        ...
                    ]
                }
                ...
             ]
             Empezá el mensaje con un "[" y terminá el mensaje con un "]"
             """

response = client.chat.completions.create(
    model="gpt-4-1106-preview",
    messages=[
        {
            "role": "system",
            "content": prompt
        },
        {
            "role": "user",
            "content": read_file(os.getcwd() + "/resources/News.txt")
        },
    ]
)

df = load_embeddings()

# parse from json


response = parse_json(response.choices[0].message.content)

for r in response:
    category = r["categoria"]
    ideas = r["ideas"]
    for index, idea in enumerate(ideas):
        r["ideas"][index] = {
            "idea": idea["idea"],
            "profesional_relacionado": search_reviews(df, idea["idea"] + ". temas:" + idea["temas"].__str__(),
                                                      n=5).tolist()
        }

print(response)
print("Time elapsed: ", time.time() - start)


find_profesional.py
import os

from openai import OpenAI

from python_dump import search_reviews, load_embeddings

from functools import reduce


def read_file(path):
    with open(path, 'r') as f:
        contents = f.read()
    return contents


key = os.environ.get('OPENAI_API_KEY')

client = OpenAI(api_key=key)

user_input = read_file(os.getcwd() + "/resources/User_prompt_find_profesional.txt")

professionals = search_reviews(load_embeddings(), user_input, 5).tolist()

print(user_input)

for r in professionals:
    content = read_file(os.getcwd() + "/resources/raw_profiles/" + r)
    prompt = """
        - A partir de la idea una columna de opinion y el perfil del profesional debes primero 
        asignar un puntaje del 1 al 10 para hablar del tema que te van a proporcionar.
        - tambien debes justificar la calificación que le diste al profesional para hablar del tema.
        - se breve y consiso.
        - Escribí Nombre:  y el nombre del profesional.
        - Escribí Calificación: y el puntaje que le diste al profesional.
        - Escribí Justificación: y la justificación que le diste al profesional.
        """
    response = client.chat.completions.create(
        model="gpt-4-1106-preview",
        messages=[
            {
                "role": "system",
                "content": prompt
            },
            {
                "role": "user",
                "content": f"""
                tema: {user_input}
                perfil: {content}
                """
            },
        ]
    )
    print(response.choices[0].message.content)



create_embeddings.py
import os

from openai import OpenAI


def iterate_files(directory, saveDirectory):
    for filename in os.listdir(directory):
        if os.path.isfile(os.path.join(directory, filename)):
            print(f"Processing file: {filename}")
            with open(os.path.join(directory, filename), 'r') as file:
                file_content = file.read()
                with open(os.path.join(saveDirectory, filename), 'w') as saveFile:
                    saveFile.write(create_embeddings(file_content).data[0].embedding.__str__())



client = OpenAI(api_key=os.environ.get('OPENAI_API_KEY'))


def create_embeddings(text):
    return client.embeddings.create(
        model="text-embedding-ada-002",
        input=[text]
    )


iterate_files(os.getcwd() + "/resources/raw_profiles", os.getcwd() + "/resources/embedded_profiles")


python_dump.py
import json
from ast import literal_eval

import numpy as np
import pandas as pd
from openai import OpenAI
import os

from myutils.embeddings_utils import get_embedding, cosine_similarity


def read_file(path):
    with open(path, 'r') as f:
        contents = f.read()
    return contents


def parse_json(string):
    return json.loads(string)


def read_file(path):
    with open(path, 'r') as f:
        contents = f.read()
    return contents


def read_embedding(file_path):
    with open(file_path, 'r') as f:
        data_str = f.read()
        data_list = literal_eval(data_str)
        return np.array(data_list)


def load_embeddings():
    directory_path = os.getcwd() + "/resources/embedded_profiles"
    data = []

    for filename in os.listdir(directory_path):
        if filename.endswith(".txt"):
            file_path = os.path.join(directory_path, filename)
            embedding = read_embedding(file_path)
            data.append((filename, embedding))

    return pd.DataFrame(data, columns=["filename", "embedding"])


def search_reviews(df, product_description, n=3, pprint=False):
    product_embedding = get_embedding(
        product_description,
        model="text-embedding-ada-002"
    )
    df["similarity"] = df.embedding.apply(lambda x: cosine_similarity(x, product_embedding))

    results = (
        df.sort_values("similarity", ascending=False)
        .head(n)
        .filename
    )
    if pprint:
        for r in results:
            print(r)
            print()
    return results


def generate_article(user_input=read_file(os.getcwd() + "/resources/User_prompt_generate_article.txt")):
    key = os.environ.get('OPENAI_API_KEY')

    client = OpenAI(api_key=key)

    prompt = """
             Vas a generar un texto en bulleteado o una columna de opinión a partir de una noticia.
             Si se especifica un perfil de un profesional, el texto debe estar escrito en ese perfil.
             El texto debe estar escrito como si lo hubiera escrito alguien con el perfil especificado si se especifica.
             El tipo de texto lo determina el usuario en el prompt. De no ser especificado, se asume que es un texto en columna de opinión.
             Si se te pasa una noticia, usala como fuente de información para generar el texto.
             Para la colmna de opinion:       
                Debe incluir un titulo breve.  
                Escribí el texto en prosa.
                A partir de la siguiente idea sobre una noticia y el contexto proporcionado del profesor y el tema generá una columna de opinión.
                La columna de opinión debe tener una introducción, un desarrollo y una conclusión.
                No escribas introduccion: desarrollo: conclusión sino que escribí el texto en prosa.
                La columna de opinión debe tener un título.
                La columna de opinión debe tener un tono formal.
             Para el bulleteado:
                A partir de la siguiente idea sobre una noticia y el contexto proporcionado del profesor y el tema generá un texto en bulleteado.
                El bulleteado es una lista de ideas que se pueden desarrollar en una columna de opinión.

            Si el usuario envia noticias, debes usarlas para contexto de la columna de opinión.
            Si no estan relacionados con la idea no los uses.
             """

    response = client.chat.completions.create(
        model="gpt-4-1106-preview",
        messages=[
            {
                "role": "system",
                "content": prompt
            },
            {
                "role": "user",
                "content": user_input
            },
        ]
    )

    print(response.choices[0].message.content)
